rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp

```diff
@@ -289,6 +290,22 @@ void SqliteStorage::write(
   commit_transaction();
 }
 
+void SqliteStorage::set_read_order(rosbag2_storage::ReadOrder read_order)
+{
+  if (read_order == read_order_) {
+    return;
+  }
+  read_order_ = read_order;
+  read_statement_ = nullptr;
+
+  if (read_order_.sort_by == rosbag2_storage::ReadOrder::PublishedTimestamp) {
+    throw std::runtime_error("Not Implemented - PublishedTimestamp read order.");
+  }
+  if (read_order.reverse && seek_row_id_ == 0) {
+    seek_row_id_ = get_last_rowid();
+  }
```

Comment:
the first ID in sqlite is 1, so 0 is an invalid row id that means the value has never been changed. When iterating backward through a file, the id will never go below 1, so I think we can't get to 0 during normal message iteration.

Reply:
@emersonknapp Though you are right that currently we can't get to 0 during normal message iteration.
But this is only because due to the another bug in `read_next()`
https://github.com/ros2/rosbag2/blob/167182a2718f8a327470e72b35e383d64044f8a5/rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp#L329-L334

We only increment `seek_row_id` and `current_message_row` but shall decrement in case of `reverse` ordering.
If we will fix this bug and start decrement on reverse. `seek_row_id` will become zero when we read out the first element from the bag.

Reply:
Ah - I've figured out why my test didn't catch this. I wasn't resetting the iterator, which is what these tracking values are for. So i need to split each ordering test into 2 - uninterrupted iteration on the same iterator, and broken reads that reset the iterator after each read, like in my other example.

Reply:
OK - I have now updated this increment (to be a conditional decrement), and the latest test caught that situation before I changed it, with the `check_against_sorted(bool reset)` signature for test data

Reply:
On further consideration - this fixes the second thing you brought up but not the original comment situation as:

* Open reader
* set read order
* seek to end
* read as far as the last message
* set read order to same value - reverse
* `has_next` should return `false` because it is still at the beginning of the file

Reply:
I am thinking now that maybe it doesn't make sense to set the row id at all on `set_read_order` - maybe the user needs to explicitly seek to the end of the file? The cases are a little bit ambiguous how it should behave.

Maybe,
* If `set_read_order(reverse=true)` set before `open`, then on `open` automatically seek to end of file
* If storage is already `open`, then `set_read_order` doesn't change the place, even if no messages have been written yet

Does that behavior make sense?

Reply:
I am thinking about it.
The problem is that we are resetting `read_statement_ = nullptr;` inside `set_read_order(order)`  and will have to readout DB with new select statement.

Reply:
@emersonknapp  it seems we need to change `seek_row_id` each time when we are toggling `reverse` flag to take it in to account in next `prepare_for_reading()` call.
```
if (read_order.reverse != read_order_.reverse) {
  seek_row_id_ = read_order_.reverse ? get_last_rowid() : 0;
}
```

Reply:
That suggestion is incorrect. That makes it so that every time you `set_read_order` (or `set_filter`, etc) you lose your place.

I'm asking for what the _behavior_ should be, independent of the code. What should be the promise of the API, when you set the read order? This is something that we state up front, before writing the code.

Which behavior should we choose? It will then be explicitly stated in the API
1. When you open a storage file, it is always set at the first message in the file. You may `set_read_order(reverse)` but this won't change your place, it is still pointing to the first message in the file (`has_next` returns false). In this case, the user must `seek` to the end of the file before starting to read messages, if they want to read from end-to-beginning.
2. If you `set_read_order(reverse)` before `open()`, then the read head will be set to the end in `open()` (`has_next` returns true directly after `open`
3. If you `set_read_order(reverse)` (before or after `open()`) before reading any messages, then the read head is set to the end of the file (`has_next` returns true)

Reply:
OK - I have decided on a slightly different approach. I am going to go with approach 1 and disable reverse-file order for now. We can leave that decision unmade. Users can `seek(start_time + duration)` to get to the end of the file, but there isn't a good way to `seek()` to the last file-order message.

Reply:
I am Ok to currently disable case when `read_order.sort_by == rosbag2_storage::ReadOrder::File && read_order.reverse)`
Although while trying to find solution for this case I found out that when we will have read order by file without reverse which is currently supported the `seek(timestamp)` will not really work and will reset our `seek_row_id` which is not expected.